WIP: mgr/devicehealth: added life_expectancy_response()

yaarith opened the pull request.

liewegas committed:
mgr/devicehealth: rename old smart module to devicehealth

Let's avoid "SMART" since it's misleading (it refers specifically to ATA).

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
mgr: enable devicehealth by default

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
osd: key smart results by unique device id

This is more useful than the current local device name.

Clean up some formatting.

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
mgr/devicehealth: add some scraping infrastructure

- command to fetch smart info
- command to scrape a device and store the metrics
- command to scrape all devices (and store)
- command to dump stored metrics
- purging of old metrics

This is based on code originally written by Yaarit.

Signed-off-by: Yaarit Hatuka yaarithatuka@gmail.com
Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
mgr/ActivePyModules: add get for 'device <devid>'

This is slightly different than the usual pattern because it is
parameterized.  I want to avoid fetching *all* devices if we don't need
it.

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
osd: optionally query smart for a single devid

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas commented:
git rebase HEAD^ --onto liewegas/wip-devicehealth

liewegas commented on src/pybind/mgr/devicehealth/module.py:
```diff
@@ -12,23 +12,33 @@
         'scrape_frequency': '86400',
         'retention_period': '86400*14',
         'pool_name': 'device_health_metrics',
+        'mark_out_threshold': 86400*14, # TODO: change back to string
```
str(86400*14)
etc

yaarith replied:
fixed

liewegas committed:
mgr/devicehealth: fix 'device get-health-metrics' when no metrics are stored

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas committed:
mgr/mgr_module: update docstring for get()

Signed-off-by: Sage Weil <sage@redhat.com>

liewegas commented on src/pybind/mgr/devicehealth/module.py:
```diff
@@ -243,6 +255,66 @@ def show_device_metrics(self, devid, sample):
+                'MGR_DEVICE_HEALTH': {
+                    'severity': 'warning',
+                    'summary': 'Imminent failure was detected on some devices.'
+                    ' Run "ceph health detail".',
```
don't need this line

yaarith replied:
fixed

liewegas commented on src/pybind/mgr/devicehealth/module.py:
```diff
@@ -243,6 +255,66 @@ def show_device_metrics(self, devid, sample):
+            self.set_health_checks({
+                'MGR_DEVICE_HEALTH': {
+                    'severity': 'warning',
+                    'summary': 'Imminent failure was detected on some devices.'
```
Imminent failure anticipated for <num> device(s)

yaarith replied:
fixed

liewegas commented on src/pybind/mgr/devicehealth/module.py:
```diff
@@ -243,6 +255,66 @@ def show_device_metrics(self, devid, sample):
+        r, outb, outs = result.wait()
+        # TODO what else should we do in case marking out failed?
+        if r != 0:
+            self.log.error('Error: Could not mark OSD %s out. r: [%s], outb: [%s], outs: [%s]' % (osd_ids, r, outb, outs))
```
log a warning, but this isn't a critical error (ceph osd out will enver really fail.. and if it does, we'll do it again the next time)

yaarith replied:
changed

Commit:
mgr/devicehealth: add helpers to life_expectancy_response()

- if mark_out_threshold is met we write to log.warn instead of raising a
  health warning.
- check that OSD is 'in' before calling mark_out().
- raise a health warning in case OSD is marked 'out' but still has PGs
  attached to it.
- cast thresholds default values to string.
- add SCSI multipath support to health warning message.
- change health warning message.

Signed-off-by: Yaarit Hatuka yaarithatuka@gmail.com

Commit:
common/blkdev remove debug statements

Signed-off-by: Yaarit Hatuka yaarithatuka@gmail.com
